atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolImpl.java

```diff
@@ -324,24 +336,84 @@ private synchronized void refreshPool() {
         absentServers.forEach(cassandraServer -> {
             CassandraClientPoolingContainer container = cassandra.removePool(cassandraServer);
             absentHostTracker.trackAbsentCassandraServer(cassandraServer, container);
         });
 
         Set<CassandraServer> serversToShutdown = absentHostTracker.incrementAbsenceAndRemove();
 
-        if (!(serversToAdd.isEmpty() && absentServers.isEmpty())) { // if we made any changes
-            sanityCheckRingConsistency();
+        Set<CassandraServer> validatedServersToAdd =
+                validateNewHostsTopologiesAndMaybeAddToPool(getCurrentPools(), serversToAdd);
+
+        if (!(validatedServersToAdd.isEmpty() && absentServers.isEmpty())) { // if we made any changes
             cassandra.refreshTokenRangesAndGetServers();
         }
 
-        logRefreshedHosts(serversToAdd, serversToShutdown, absentServers);
+        Preconditions.checkState(
+                !getCurrentPools().isEmpty() || serversToAdd.isEmpty(),
+                "No servers were successfully added to the pool. This means we could not come to a consensus on"
+                    + " cluster topology, and the client cannot start as there are no valid hosts. This state should"
+                    + " be transient (<5 minutes), and if it is not, indicates that the user may have accidentally"
+                    + " configured AltasDB to use two separate Cassandra clusters (i.e., user-led split brain).",
+                SafeArg.of("serversToAdd", serversToAdd));
+
+        logRefreshedHosts(validatedServersToAdd, serversToShutdown, absentServers);
+    }
+
+    /**
+     * Validates new servers to add to the cassandra client container pool,
+     * by checking them with the @see {@link com.palantir.atlasdb.keyvalue.cassandra.CassandraTopologyValidator}.
+     * If any servers come back and are not in consensus this is OK, we will simply add them to the absent host
+     * tracker, as we most likely will retry this host in subsequent calls.
+     *
+     * @return The set of cassandra servers which have valid topologies and have been added to the pool.
+     */
+    @VisibleForTesting
+    Set<CassandraServer> validateNewHostsTopologiesAndMaybeAddToPool(
+            Map<CassandraServer, CassandraClientPoolingContainer> currentContainers,
+            Set<CassandraServer> serversToAdd) {
+        if (serversToAdd.isEmpty()) {
+            return Set.of();
+        }
+
+        Map<CassandraServer, CassandraClientPoolingContainer> serversToAddContainers =
+                getContainerForNewServers(serversToAdd);
+
+        Map<CassandraServer, CassandraClientPoolingContainer> allContainers =
+                ImmutableMap.<CassandraServer, CassandraClientPoolingContainer>builder()
+                        .putAll(serversToAddContainers)
+                        .putAll(currentContainers)
+                        .buildOrThrow();
```

Comment:
(Seems like the comment has disappeared, but this was previously buildKeepingLast. For posterity, 
serversToAdd are servers that are not in the current server list.
If they were ever in the server list, then they are either tracked by the absent host tracker (in which case we reuse the pool), or we've shutdown the pool and need to recreate it.
If they are in the absent host tracker, then they've either been removed from the pool (setServersInPool) or never added (just below this).

In essence, we didn't need the whole merge map things in the end :)